negotiation/sessions/hybrid_session.py

```diff
@@ -0,0 +1,353 @@
+        self.bottomline = None
+        self.listing_price = self.kb.listing_price
+        self.category = self.kb.category
+        self.inc = None
+
+    def choose_action(self):
+        # convert lf tokens into int
+        lfs = [map(self.mappings['lf_vocab'].to_ind, lf) for lf in self.lf_tokens]
+        batch = self.create_batch(lfs)
+        encoder_init_state = None
+        output_data = self.generator.generate_batch(batch)
+        entity_tokens = self._output_to_tokens(output_data)
+        action = entity_tokens[1] # first element is just start symbol
+
+        return action
+
+    '''
+    action = self.manager.choose_action(state=self.state)
+    if not action:
+        action = self.retrieve_action()
+        if not action in self.manager.available_actions(self.state):
+            action = 'unknown'
+    return action
+
+    def retrieve_action(self):
+        template = self.retrieve_response_template(None)
+        action = template['tag']
+        print 'retrieved action:', action
+        return action
+    '''
+
+    def shorten_title(self, title):
+        """If the title is too long, shorten it using a generic name for filling in the templates.
+        """
+        if len(title.split()) > 3:
+            if self.kb.category == 'car':
+                return 'car'
+            elif self.kb.category == 'housing':
+                return 'apartment'
+            elif self.kb.category == 'phone':
+                return 'phone'
+            elif self.kb.category == 'bike':
+                return 'bike'
+            else:
+                return 'item'
+        else:
+            return title
+
+    @classmethod
+    def get_fraction(cls, zero, one, fraction):
+        # see rulebased session for details
+        return one * fraction + zero * (1. - fraction)
+
+    def round_price(self, price):
+        """Round the price so that it's not too specific.
+        """
+        if price == self.target:
+            return price
+        if price > 100:
+            return int(round(price, -1))
+        if price > 1000:
+            return int(round(price, -2))
+        return int(round(price))
+
+    def estimate_bottomline(self):
+        raise NotImplementedError
+
+    def init_price(self):
+        """Initial offer.
+        """
+        raise NotImplementedError
+
+    def receive(self, event):
+        # returns an Utterance object with LF attribute
+        utterance = self.parser.parse(event, self.state)
+        # returns a list of tokens which masquerades as an utterance
+        # utterance = self.env.preprocessor.process_event(event, self.kb)
+        print 'receive:', utterance
+
+        self.state.update(self.partner, utterance)
+        intent = self.process_lf(self.agent)
+        self.lf_tokens.append(intent)
+
+        if self.bottomline is None:
+            self.bottomline = self.estimate_bottomline()
+        # self.dialogue.add_utterance(event.agent, utterance)
+
+    def process_lf(self, agent):
```

Comment:
use lf_to_tokens? these preprocess should be in preprocessor and reused as much as possible to make sure it's consistent for both training and testing